Add PauliLindbladMap.parity_sample method

DanPuzzuoli opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Adds the method `PauliLindbladMap.parity_sample`.

This method is very similar to `PauliLindbladMap.signed_sample`, however:
- The sign convention is flipped. `signed_sample` associates `True` with `+1` and `False` with `-1`, which is the opposite of the expected convention.
- `parity_sample` adds the optional arguments `scale` and `local_scale` that allows the user to scale the rates before sampling *without* modifying the rates stored in the instance (or creating a whole new instance).

### Details and comments

In terms of code, this PR:
- Modifies the underlying rust sampling method, renaming it to `parity_sample`, and adding handling for the optional `scale` and `local_scale` methods.
- Changes `PauliLindbladMap.signed_sample` and `PauliLindbladMap.sample` to use the new method, passing `None` to the optional arguments.
- Adds the `PauliLindbladMap.parity_sample` python/rust interfacing method.
- Adds tests for `parity_sample` similar to those for `signed_sample`.

For documentation:
- The error message in `PauliLindbladMap.sample` that is raised if any rates are negative, is now changed to suggest using `PauliLindbladMap.parity_sample` instead of `PauliLindbladMap.signed_sample`.
- The doc string of `PauliLindbladMap.signed_sample` now opens with a statement that `parity_sample` is the preferred method, with `signed_sample` having a non-standard convention.

For reviewers:
- I'm not sure how the comment that `parity_sample` is preferred over `signed_sample` should be formatted.
- At the beginning of the rust `parity_sample` method, the values of the variables `probabilities` and `non_negative_rates` are set with a `match` block. The behaviour that I wanted is: if `scale` and `local_scale` are both `None`, then the pre-existing values in `self.probabilities` and `self.non_negative_values` should be used. If either is not `None`, then it's necessary to compute `probabilities` and `non_negative_rates` from a modified version of the rates. I however ran into an issue: in the first case I want to borrow `self.probabilities` and `self.non_negative_values`, but then that becomes inconsistent with the second case (which I can't borrow because the values calculated in that tbranch will disappear if not assigned). To get it to work I ended up cloning `self.probabilities` and `self.non_negative_values` as a hack, but ideally we wouldn't be cloning this.

DanPuzzuoli committed:
initial implementation

DanPuzzuoli committed:
adding tests

DanPuzzuoli committed:
adding documentation

DanPuzzuoli committed:
formatting

DanPuzzuoli committed:
deleting old rust sample method

DanPuzzuoli committed:
adding release note

DanPuzzuoli committed:
updating release note

ihincks commented on releasenotes/notes/pauli-lindblad-map-parity-sample-cbb1ed09556a24db.yaml:
```diff
@@ -0,0 +1,8 @@
+    Adds the :meth:`.PauliLindbladMap.parity_sample` method. This method is very similar to the
+    pre-existing :meth:`.PauliLindbladMap.signed_sample` method, however it uses a sign convention
+    that is more consistent with the rest of qiskit, and therefore is the preferred method to use
+    goign forward. It has additional optional arguments to apply scalings to the rates when sampling
```
```suggestion
    going forward. It has additional optional arguments to apply scalings to the rates when sampling
```

aeddins-ibm commented on test/python/quantum_info/test_pauli_lindblad_map.py:
```diff
@@ -1231,6 +1231,139 @@ def test_signed_sample(self):
+            "X": probs[0] * (1 - probs[1]),
+            "Y": (1 - probs[0]) * probs[1],
+            "Z": (1 - probs[0]) * (1 - probs[1]),
+        }
```
In this test (and many tests in this file), are the probabilities for `X` and `Y` errors accidentally swapped? Should it be like this instead (since it looks like the chosen convention is to work with probabilities of each error NOT happening)?

```suggestion
        pauli_lindblad_map = PauliLindbladMap([("X", -1.0), ("Y", -1.0)])
        probs = pauli_lindblad_map.probabilities()
        probs_dict = {
            "I": probs[0] * probs[1],
            "X": (1-probs[0]) * probs[1],
            "Y": probs[0] * (1-probs[1]),
            "Z": (1 - probs[0]) * (1 - probs[1]),
        }
```

DanPuzzuoli replied:
Great catch on this. This was initially very concerning but it turns out I wrote every single test case with `probs[0] == probs[1]`, so my incorrect formulas are still producing the right values.

I'll implement this fix and also change the values a bit so that this would have failed with the incorrect formulas.

debasmita2102 replied:
For the Python test code around lines 1240–1250, may be you can consider adding a clear docstring specifying the intended error probability convention, parameterizing the test to check both canonical and swapped orderings, and expanding assertion messages for clarity. This can catch subtle bugs and ensure test robustness. Additionally, helper functions for probability indexing would help maintain consistency across the test suite, i guess.

aeddins-ibm commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -384,24 +384,56 @@ impl PauliLindbladMap {
     /// Sample sign and Pauli operator pairs from the map.
-    pub fn sample(&self, num_samples: u64, seed: Option<u64>) -> (Vec<bool>, QubitSparsePauliList) {
+    /// Note that here the "sign" bool is interpreted as the exponent of (-1)^b, which is the
+    /// opposite convention to the sample function.
```
I wasn't sure, should this be changed from "sample function" to "signed_sample function"?
```suggestion
    /// opposite convention to the signed_sample function.
```

DanPuzzuoli replied:
Actually this is in the rust class and is referencing the rust-class `PauliLindbladMap.sample` method, which no longer exists, there is only `parity_sample` now. I'm going to remove this statement.

aeddins-ibm commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -384,24 +384,56 @@ impl PauliLindbladMap {
+                        .iter()
+                        .zip(local_scale.clone())
+                        .map(|(r, s)| r * s)
+                        .collect(),
```
I don't know Rust that well, would these be simpler as `if` statements?

DanPuzzuoli replied:
Yeah - how to do this in the right rust-y way is something I'd like input on as well. This code is a result of my confusion around conditionally assigning variable values in rust :P. My performance-related concern with this is that I'm cloning the vectors in the `(None, None)` case, which is undesirable.

debasmita2102 replied:
On lines 401–417, I notice the match statement can end up cloning vectors (for example, when both scale and local_scale are None). Would it make sense to investigate alternatives like borrowing (&self.probabilities) or using Cow (clone on write) to avoid unnecessary heap allocations? In data-heavy workloads, repeated cloning may increase memory usage and may lead to slower performance, especially when sampling many times inside loops. If the vectors are large or sampled often, avoiding deep clone operations could make this code more scalable.

DanPuzzuoli replied:
Yes this is one of the things I was wanting input on. I think this should be changed but I didn't know how to do it properly in rust.

jakelishman replied:
I've showed one potential path for doing this in 4957f28149a7dbefebc48e398162bc47319a39a3.  I don't think I necessarily got the absolute neatest way possible, but I'm in a rush haha.

aeddins-ibm commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -1577,12 +1664,12 @@ impl PyPauliLindbladMap {
         for non_negative in inner.non_negative_rates.iter() {
             if !non_negative {
                 return Err(PyValueError::new_err(
-                    "PauliLindbladMap.sample called for a map with negative rates. Use PauliLindbladMap.signed_sample",
+                    "PauliLindbladMap.sample called for a map with negative rates. Use PauliLindbladMap.parity_sample",
                 ));
             }
         }
 
-        let (_, paulis) = py.detach(|| inner.sample(num_samples, seed));
+        let (_, paulis) = py.detach(|| inner.parity_sample(num_samples, seed, None, None));
 
         paulis.into_pyobject(py)
     }
```
Not essential for this PR, but long term should we delete this method? It doesn't seem to add much.

Perhaps we can eventually get rid of `signed_sample` and `parity_sample`, and just have `sample` that returns the Paulis unless given an optional arg `return_signs = True`?

DanPuzzuoli replied:
We could yes, though in the planning discussion we had before I wrote this PR, it was more-or-less concluded that it's not necessarily worth going through the deprecation process so long as the effect of these methods are all consistently documented.

DanPuzzuoli replied:
I think maybe if it is ever confusing to users then it'd be worth it, but for now I'm partial to just leaving them as is, so long as the preferred method is obvious.

debasmita2102 replied:
I agree with @aeddins-ibm With several similar sampling methods (sample, signed_sample, parity_sample) implemented around lines 1667–1680, would combining them into a single function—perhaps with a flag or enum to control sign conventions—be considered? For example, a function like sample(sign_mode, ...) or keyword arguments (in Python or Rust) lets users pick the behavior explicitly. This helps avoid confusion about which method is recommended, enables easier documentation, and reduces the risk that future updates only reach one variant.

ihincks replied:
The choice to add a new method rather than add a selector argument to the existing method is specifically to fix the problem that the current implementation returns the wrong convention---we'd require the default value of the selector to result in the wrong convention to avoid a breaking change, and that doesn't seem adequately fix the problem from my perspective.

I'd be happy to mark `signed_sample` as deprecated, though @jakelishman has mentioned elsewhere that he doesn't see it as pressing, IIRC.

jakelishman replied:
We can't deprecate `signed_sample` until the release after `parity_sample` (its alternative) is released, otherwise there's no warningless upgrade path for users.

aeddins-ibm reviewed (commented):
Basically LGTM, thanks for switching this convention.

Minor comments
- Likely typo with `parity_sample` docstring referring to `sample` vs `signed_sample`
- Looks like there are indexing errors in the tests, where they don't matter for the tests but could confuse someone trying to learn/generalize from the tests.
- Long term can we move to just one `sample` method instead of 2 or 3?

DanPuzzuoli committed:
Update releasenotes/notes/pauli-lindblad-map-parity-sample-cbb1ed09556a24db.yaml

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

DanPuzzuoli committed:
fixing documentation for rust parity_sample method

DanPuzzuoli committed:
fixing tests

jakelishman committed:
Merge remote-tracking branch 'ibm/main' into parity-sample

jakelishman committed:
Remove cloning when no mutation is necessary

The trick is assigning a _reference_ rather than a `Vec`, but declaring
the storage space for an intermediate `Vec` in the outer scope to extend
its lifetime.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -384,24 +384,52 @@ impl PauliLindbladMap {
+        num_samples: u64,
+        seed: Option<u64>,
+        scale: Option<f64>,
+        local_scale: Option<Vec<f64>>,
```
We might want to consider taking this as a Numpy array-like in the future, so there's a path to cloning-free local rates (a performance improvement), but we can do that without breaking the Python API, so no worries now.

jakelishman reviewed (approved):
I've addressed the cloning comment from above for others, and merged this up with `main`.  It generally looks ok to me, but I'm mostly trusting the reviews from Ian, Andrew and Debasmita.
